test_fms/axis_utils/test_axis_utils.F90

```diff
@@ -17,150 +17,775 @@
+    & special_axis_names_y = [character(13) :: "lat", "y", "degrees_n", "degrees_north", "degreesn"], &
+    & special_axis_names_z = [character(6) :: "depth", "height", "z", "cm", "m", "pa", "hpa"], &
+    & special_axis_names_t = [character(4) :: "time", "t", "sec", "min", "hou", "day", "mon", "yea"], &
+    & attr_names           = [character(14) :: "cartesian_axis", "axis"], &
+    & xyzt_uc              = ["X", "Y", "Z", "T"]
+
+  call open_netcdf_w(test%fileobj)
+  call register_axis(test%fileobj, "dim1", 1)
+
+  ! Check a variable which does not exist
+
+  allocate(test_nonexistent_var)
+  test_nonexistent_var%var = "does_not_exist"
+  test_nonexistent_var%cart = "N"
+
+  test%test0 => test_nonexistent_var
+  test%test1 => test_nonexistent_var
+
+  ! Check a variable which exists, but which has neither a "cartesian_axis" nor an "axis" attribute.
+  var_name = "exists_no_attributes"
+  call get_axis_cart_test_add(test, var_name, "N")
+
+  do i=1,size(attr_names)
+    attr_name = trim(attr_names(i))
+
+    ! Check an unknown value on a "cartesian_axis" or "axis" attribute
+    ! TODO: This test fails. Should get_axis_cart be changed, or should this test be changed?
+    attr_value = "unexpected"
+    var_name = attr_name // "_attr_value_" // attr_value
+    call get_axis_cart_test_add(test, var_name, "N")
+    call register_variable_attribute(test%fileobj, var_name, attr_name, attr_value, str_len=len(attr_value))
+
+    do j=1,size(xyzt_uc)
+      ! Check upper-case "axis" attributes"
+      attr_value = xyzt_uc(j)
+      var_name = attr_name // "_attr_value_" // attr_value
+      call get_axis_cart_test_add(test, var_name, xyzt_uc(j))
+      call register_variable_attribute(test%fileobj, var_name, attr_name, attr_value, str_len=len(attr_value))
+
+      ! Check lower-case "axis" attributes"
+      attr_value = lowercase(xyzt_uc(j))
+      var_name = attr_name // "_attr_value_" // attr_value
+      call get_axis_cart_test_add(test, var_name, xyzt_uc(j))
+      call register_variable_attribute(test%fileobj, var_name, attr_name, attr_value, str_len=len(attr_value))
+    enddo
+  enddo
+
+  call test_special_axis_names(test, special_axis_names_x, "X")
+  call test_special_axis_names(test, special_axis_names_y, "Y")
+  call test_special_axis_names(test, special_axis_names_z, "Z")
+  call test_special_axis_names(test, special_axis_names_t, "T")
+
+  call close_file(test%fileobj)
+
+  call get_axis_cart_tests_run(test)
+end subroutine
 
-do i=1,10
-   data_in%var(i) = real(i, kind=r8_kind)-0.5_r8_kind
+subroutine get_axis_cart_test_add(test, var_name, cart)
+  type(GetAxisCartTest_t), intent(inout) :: test
+  type(GetAxisCartTestCase_t), pointer :: test_case
+  character(*), intent(in) :: var_name
+  character(1), intent(in) :: cart
 
-   data_in%var_edges(1,i) = real(i-1, kind=r8_kind)
-   data_in%var_edges(2,i) = real(i, kind=r8_kind)
+#define r4_kind "float"
```

Comment:
It replaces those definitions with the strings `"float"` and `"double"`, but only for line 210, where a string representation of the kind is needed. Both macros are `undef`'d immediately afterward.

Reply:
Explanatory comment added in 682924fecc44e8508accd542c25d547fa093fb92.

Reply:
Can you explain this more - specifically, what is actually happening with the character def and the existing AU_TEST_KIND_ macro?

Reply:
Explain it in more detail here, or in the comment in the code?

In the preprocessor's first round of substitution, `AU_TEST_KIND_` gets replaced with, e.g., `r4_kind`. Then in the second round of substitution, `r4_kind` gets replaced with `"float"` due to the very-short-lived `r4_kind` macro. This allows us to get a string representation of the kind which is suitable for passing to `register_field`.

Reply:
Thanks for explaining here and it's what I thought was happening.  Is there a reason to not simply define a second macro to substitute in _float_ when AU_TEST_KIND==_r4_kind_ and _double_ when AU_TEST_KIND==_r8_kind_?

Reply:
The preprocessor can't do string comparison, so that would only be possible via `#if`/`#else` directives if `AU_TEST_KIND_` were to take on the values `4` and `8` (rather than `r4_kind` and `r8_kind`). Another potential option is to define a second macro in `Makefile.am` that takes on values `"float"` and `"double"`, but my personal preference is to avoid offloading additional definitions to `Makefile.am` when it's possible to keep them contained in the source file.

Reply:
since k is a global variable,  can somethiing like 
 if( kind(k)=r4_kind) call register_field(test%fileobj, var_name, "float", dimensions=["dim1"]) be used instead of using macros?

Reply:
Macros replaced with runtime `if/else` in 4e1434f9314a87bef1a52cb13ae57ea05a75ae29.